File: src/utils/token_tracker.py
import os
from datetime import datetime

TRACKER_FILE = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), "tracker.txt")

def log_token_usage(agent_name: str, response: any):
    """
    Extracts usage_metadata from an AIMessage and appends it to tracker.txt.
    """
    try:
        # Some structured outputs wrap the original message, or the message itself has usage_metadata
        # Try to find usage_metadata
        metadata = None
        if hasattr(response, "usage_metadata"):
            metadata = response.usage_metadata
        elif isinstance(response, dict) and "usage_metadata" in response:
            metadata = response["usage_metadata"]
            
        if metadata:
            input_tokens = metadata.get("input_tokens", 0)
            output_tokens = metadata.get("output_tokens", 0)
            total_tokens = metadata.get("total_tokens", 0)
            
            log_line = f"[{datetime.now().isoformat()}] {agent_name} | Input: {input_tokens} | Output: {output_tokens} | Total: {total_tokens}\n"
            
            with open(TRACKER_FILE, "a") as f:
                f.write(log_line)
    except Exception as e:
        # Silently fail if tracker has issues so we don't break the pipeline
        pass

File: src/utils/token_tracker.ts
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// tracker.txt is in the workspace root directory
const TRACKER_FILE = path.join(__dirname, "..", "..", "tracker.txt");

/**
 * Extracts usage_metadata from an LLM response and appends it to tracker.txt.
 */
export function logTokenUsage(agentName: string, response: any): void {
  try {
    let metadata: any = null;

    if (response && typeof response === "object") {
      if ("usage_metadata" in response) {
        metadata = response.usage_metadata;
      } else if (response.response_metadata && response.response_metadata.tokenUsage) {
        // OpenAI LangChain format helper
        const usage = response.response_metadata.tokenUsage;
        metadata = {
          input_tokens: usage.promptTokens || 0,
          output_tokens: usage.completionTokens || 0,
          total_tokens: usage.totalTokens || 0,
        };
      } else if (response.usage) {
        metadata = {
          input_tokens: response.usage.prompt_tokens || 0,
          output_tokens: response.usage.completion_tokens || 0,
          total_tokens: response.usage.total_tokens || 0,
        };
      }
    }

    if (metadata) {
      const inputTokens = metadata.input_tokens || 0;
      const outputTokens = metadata.output_tokens || 0;
      const totalTokens = metadata.total_tokens || 0;

      const logLine = `[${new Date().toISOString()}] ${agentName} | Input: ${inputTokens} | Output: ${outputTokens} | Total: ${totalTokens}\n`;

      fs.appendFileSync(TRACKER_FILE, logLine);
    }
  } catch (error) {
    // Silently fail to not interrupt execution
  }
}
